hlc: remove SetMaxOffset

tamird opened the pull request:
...and rewrite countless tests that depend on it, and generally have
bad clock discipline. In particular, many tests previously assumed that
the clock they use starts at zero, leading to sloppy assertions at best
and completely bogus tests at worst. See
storage.TestReplicaTSCacheLowWaterOnLease for an example of the latter.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10376)
<!-- Reviewable:end -->

tamird committed:
hlc: remove SetMaxOffset

...and rewrite countless tests that depend on it, and generally have
bad clock discipline. In particular, many tests previously assumed that
the clock they use starts at zero, leading to sloppy assertions at best
and completely bogus tests at worst. See
storage.TestReplicaTSCacheLowWaterOnLease for an example of the latter.

tamird committed:
rpc,util/hlc: forbid trivial clocks

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 59 of 59 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_test.go, line 3584 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYyiZHOw87QDSrA2IS:-KVYyiZHOw87QDSrA2IT:bokltnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c630230dbfa927250a50d29653dcfaf8c528f69d/pkg/storage/replica_test.go#L3584)):*
> ```Go
> 
> 		pusher := newTransaction("test", key, 1, enginepb.SERIALIZABLE, tc.Clock())
> 		pushee := newTransaction("test", key, 1, iso, tc.Clock())
> ```

You didn't have to interchange `pusher` and `pushee`, right? Just making sure since I think this test also has a diff in #10327.

---

*[pkg/util/hlc/hlc.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYzWDkkA9-d2_O1EB1:-KVYzWDkkA9-d2_O1EB2:b-qf7kew) ([raw file](https://github.com/cockroachdb/cockroach/blob/c630230dbfa927250a50d29653dcfaf8c528f69d/pkg/util/hlc/hlc.go#L162)):*
> ```Go
> 	}
> 	ts := c.mu.timestamp
> 	c.mu.Unlock()
> ```

no convincing reason not to keep this `defer`ed in sight.

---

*[pkg/util/hlc/hlc.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYzo3CtqFp7P5k0XcK:-KVYzo3CtqFp7P5k0XcL:bvuga7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/util/hlc/hlc.go#L84)):*
> ```Go
> func NewManualClock(nanos int64) *ManualClock {
> 	if nanos == 0 {
> 		panic("zero clock is verboten")
> ```

s/verboten/forbidden/
(or at least you're the only person I know who actually uses that word)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-:-KVYzwb9OA4cJyZFntLu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 57 of 59 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/cluster/localcluster.go, line 775 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYy9BIDAeLsMD7vBT9:-KVYy9BIDAeLsMD7vBTA:b-1dgdok) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/acceptance/cluster/localcluster.go#L775)):*
> ```Go
> 		SSLCert:    filepath.Join(l.CertsDir, security.EmbeddedNodeCert),
> 		SSLCertKey: filepath.Join(l.CertsDir, security.EmbeddedNodeKey),
> 	}, hlc.NewClock(hlc.UnixNano, 0), stopper).GRPCDial(l.Nodes[i].Addr(DefaultTCP).String())
> ```

This very long expression is hard to read; it's better to keep the rpcContext local variable and call GRPCDial on it in a separate statement. 

---

*[pkg/cli/start.go, line 462 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYz3AP9I0Ttmf-nfsy:-KVYz3AQoNgZcWP-0mgv:bgh2qj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/cli/start.go#L462)):*
> ```Go
> 	stopper := stop.NewStopper()
> 	rpcContext := rpc.NewContext(
> 		log.AmbientContext{}, serverCfg.Config, hlc.NewClock(hlc.UnixNano, 0), stopper,
> ```

Most of the time you use time.Nanosecond instead of 0 for the maxOffset. Why the difference?

---

*[pkg/internal/client/rpc_sender.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYy6PwQ-SW0nuORgUM:-KVYy6PwQ-SW0nuORgUN:bv3k2qe) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/internal/client/rpc_sender.go#L36)):*
> ```Go
> // This must not be used by server.Server or any of its components, only by
> // clients talking to a Cockroach cluster through the external interface.
> func NewSender(conn *grpc.ClientConn) Sender {
> ```

Why this change? 

---

*[pkg/server/node.go, line 180 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVZ-wcGEZ-ZajEU_iQr:-KVZ-wcGEZ-ZajEU_iQs:b-fhcgid) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/server/node.go#L180)):*
> ```Go
> 	cfg.MetricsSampleInterval = time.Duration(math.MaxInt64)
> 	cfg.ConsistencyCheckInterval = 10 * time.Minute
> 	cfg.Clock = hlc.NewClock(hlc.UnixNano, 50*time.Millisecond)
> ```

Where did this value come from?

---

*[pkg/sql/kv_test.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVZ04B1xIThlGgoKbwL:-KVZ04B1xIThlGgoKbwM:bi3622f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/sql/kv_test.go#L69)):*
> ```Go
> 		SSLCert:    filepath.Join(security.EmbeddedCertsDir, "node.crt"),
> 		SSLCertKey: filepath.Join(security.EmbeddedCertsDir, "node.key"),
> 	}, nil, s.Stopper()).GRPCDial(s.ServingAddr())
> ```

Put the GRPCDial call in a separate statement.

---

*[pkg/storage/client_raft_test.go, line 2355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVZ0nXz1GTpdKMU7DQD:-KVZ0nXz1GTpdKMU7DQE:b-osn38f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/storage/client_raft_test.go#L2355)):*
> ```Go
> 	defer mtc.Stop()
> 
> 	// TODO(tamird): bad assumption?
> ```

Should be safe. The range initially appears only on store 0, which must acquire the lease to be able to replicate the range onto store 1 below. 

---

*[pkg/testutils/localtestcluster/local_test_cluster.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYxLsCAOs_PRR5fNV8:-KVYxLsCAOs_PRR5fNV9:b-x9zquq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/testutils/localtestcluster/local_test_cluster.go#L113)):*
> ```Go
> 	ltc.DB = client.NewDBWithContext(ltc.Sender, *ltc.DBContext)
> 	transport := storage.NewDummyRaftTransport()
> 	cfg, _ := storage.TestStoreConfig()
> ```

It's ugly to construct a clock inside TestStoreConfig then throw it away and replace it. This code would need to be updated if TestStoreConfig later does something else with its clock in addition to setting `cfg.Clock`. I think it would be better to let a clock be passed in to TestStoreConfig instead of returning one. 

---

*[pkg/util/hlc/hlc.go, line 53 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYvEO9YdJ-wNNn8uVh:-KVYvEO9YdJ-wNNn8uVi:b-9qkhku) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/util/hlc/hlc.go#L53)):*
> ```Go
> 	physicalClock func() int64
> 
> 	// The maximal offset of the physical clock from a quorum of known nodes. If
> ```

This value is used for two separate things: the maximum offset between the HLC's walltime portion from this node's physical clock (this is enforced by the HLC itself) and the maximum offset between this node's clock and some view of the cluster's clocks (which is enforced separately and only coincidentally uses the HLC's MaxOffset). The comments that were on SetMaxOffset are still valid and should be restored. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-:-KVZ46vFU_W55h8ucRbm:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 50 of 59 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/acceptance/cluster/localcluster.go, line 775 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYy9BIDAeLsMD7vBT9:-KVZWEyD7g1UAv0_yDgv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/acceptance/cluster/localcluster.go#L775)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This very long expression is hard to read; it's better to keep the rpcContext local variable and call GRPCDial on it in a separate statement. 
</details>
Done.

---

*[pkg/cli/start.go, line 462 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYz3AP9I0Ttmf-nfsy:-KVZUjMii3hcKpl_TiZW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/cli/start.go#L462)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Most of the time you use time.Nanosecond instead of 0 for the maxOffset. Why the difference?
</details>
Done.

---

*[pkg/internal/client/rpc_sender.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYy6PwQ-SW0nuORgUM:-KVZUlIdZ4aQX0nVQchJ:b-faa96g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/internal/client/rpc_sender.go#L36)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why this change? 
</details>
This change allowed me to rewrite `cli.makeDBClient` in terms of `cli.getGRPCConn`. Those methods previously duplicated lots of code.

---

*[pkg/server/node.go, line 180 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVZ-wcGEZ-ZajEU_iQr:-KVZVIiEdkp3HOl2vZRE:bf3xuug) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/server/node.go#L180)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Where did this value come from?
</details>
Good question. Changed to time.Nanosecond.

---

*[pkg/sql/kv_test.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVZ04B1xIThlGgoKbwL:-KVZVYylIYQT-vayS5PD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/sql/kv_test.go#L69)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Put the GRPCDial call in a separate statement.
</details>
Done.

---

*[pkg/storage/client_raft_test.go, line 2355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVZ0nXz1GTpdKMU7DQD:-KVZVc6yVfBICqdfF2A7:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/storage/client_raft_test.go#L2355)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Should be safe. The range initially appears only on store 0, which must acquire the lease to be able to replicate the range onto store 1 below. 
</details>
Removed.

---

*[pkg/storage/replica_test.go, line 3584 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYyiZHOw87QDSrA2IS:-KVZTe4rzEfKPYmXczev:b-paifyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c630230dbfa927250a50d29653dcfaf8c528f69d/pkg/storage/replica_test.go#L3584)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You didn't have to interchange `pusher` and `pushee`, right? Just making sure since I think this test also has a diff in #10327.
</details>
I did; `newTrasaction` calls `.Now()` on the clock, which moves the HLC, so the other order didn't work.

---

*[pkg/util/hlc/hlc.go, line 162 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYzWDkkA9-d2_O1EB1:-KVZVolFzxqOkNwVtE7d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c630230dbfa927250a50d29653dcfaf8c528f69d/pkg/util/hlc/hlc.go#L162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> no convincing reason not to keep this `defer`ed in sight.
</details>
Done.

---

*[pkg/util/hlc/hlc.go, line 53 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYvEO9YdJ-wNNn8uVh:-KVZWoFcpezBDxnGUmHN:b21paww) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/util/hlc/hlc.go#L53)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This value is used for two separate things: the maximum offset between the HLC's walltime portion from this node's physical clock (this is enforced by the HLC itself) and the maximum offset between this node's clock and some view of the cluster's clocks (which is enforced separately and only coincidentally uses the HLC's MaxOffset). The comments that were on SetMaxOffset are still valid and should be restored. 
</details>
The comments on SetMaxOffset talk about an "offset from the cluster", which is not really a thing - that's why I didn't just dump that comment here verbatim.

I've added the bit about the HLC walltime.

---

*[pkg/util/hlc/hlc.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYzo3CtqFp7P5k0XcK:-KVZYIQhUdrxAf4-E6d5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/util/hlc/hlc.go#L84)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/verboten/forbidden/
> (or at least you're the only person I know who actually uses that word)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 50 of 59 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/testutils/localtestcluster/local_test_cluster.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYxLsCAOs_PRR5fNV8:-KVZWavD4f0NCRAU5OL7:bzctgam) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/testutils/localtestcluster/local_test_cluster.go#L113)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's ugly to construct a clock inside TestStoreConfig then throw it away and replace it. This code would need to be updated if TestStoreConfig later does something else with its clock in addition to setting `cfg.Clock`. I think it would be better to let a clock be passed in to TestStoreConfig instead of returning one. 
</details>
Agreed, but not done. In practice, there are quite a few cases where clocks are overwritten and we should address that in some comprehensive way. In particular, `mtc`'s handling of clocks is still quite sloppy and changing TestStoreConfig to take a clock doesn't significantly change that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 45 of 60 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/testutils/localtestcluster/local_test_cluster.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376#-KVYxLsCAOs_PRR5fNV8:-KVZd_Sff89fjpLsMD-c:b-58f5yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a48469ad5327bac923b0bfc99f27dd3ce66f3486/pkg/testutils/localtestcluster/local_test_cluster.go#L113)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Agreed, but not done. In practice, there are quite a few cases where clocks are overwritten and we should address that in some comprehensive way. In particular, `mtc`'s handling of clocks is still quite sloppy and changing TestStoreConfig to take a clock doesn't significantly change that.
</details>
Pushed a commit that does this, not clear the benefit is worth it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10376)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I was thinking of removing the newly-added second return value from TestStoreConfig, which could undo a lot of `, _` changes introduced in this PR. Tests that need a manual clock would make their own and pass it in (or use a new helper function, instead of changing TestStoreConfig which is used in many places that don't care about the underlying clock).

tamird commented:
@bdarnell done, PTAL

tamird committed:
storage: remove reservationTimeout from test

Also use math.MaxInt32 instead of time.Hour to remove ambiguity.

Note that time.Duration is an alias of int64, but there are other
intervals in the system on which unchecked arithmetic is performed, so
MaxInt32 is safer than MaxInt64, and provides a similar benefit.
